grafana_collector/report/report.go

```diff
@@ -0,0 +1,245 @@
+	return new(g, dashName, time)
+}
+
+func new(g grafana.Client, dashName string, time grafana.TimeRange) *report {
+	tmpDir := filepath.Join("tmp", uuid.New())
+	return &report{g, time, dashName, tmpDir}
+}
+
+// Generate returns the report.pdf file. After reading this file it should be Closed()
+// After closing the file, call report.Clean() to delete the file
+func (rep *report) Generate() (pdf io.ReadCloser, err error) {
+	dash, err := rep.gClient.GetDashboard(rep.dashName)
+	if err != nil {
+		err = fmt.Errorf("error fetching dashboard %v: %v", rep.dashName, err)
+		return
+	}
+	err = rep.renderPNGsParallel(dash)
+	if err != nil {
+		err = fmt.Errorf("error rendering PNGs in parralel for dash %+v: %v", dash, err)
+		return
+	}
+	pdf, err = rep.renderPDF(dash)
+	return
+}
+
+// Clean deletes the temporary directory used during report generation
+func (rep *report) Clean() {
+	err := os.RemoveAll(rep.tmpDir)
+	if err != nil {
+		log.Errorf("Error cleaning up tmp dir: %v", err)
+	}
+}
+
+func (rep *report) imgDirPath() string {
+	return filepath.Join(rep.tmpDir, imgDir)
+}
+
+func (rep *report) pdfPath() string {
+	return filepath.Join(rep.tmpDir, reportPdf)
+}
+
+func (rep *report) renderPNGsParallel(dash grafana.Dashboard) error {
+	//buffer all panels on a channel
+	panels := make(chan grafana.Panel, len(dash.Panels))
+	for _, p := range dash.Panels {
+		panels <- p
+	}
+	close(panels)
+
+	//fetch images in parrallel form Grafana sever.
+	//limit concurrency using a worker pool to avoid overwhelming grafana
+	//for dashboards with many panels.
+	var wg sync.WaitGroup
+	workers := 5
+	wg.Add(workers)
+	errs := make(chan error, len(dash.Panels)) //routines can return errors on a channel
+	for i := 0; i < workers; i++ {
+		go func(panels <-chan grafana.Panel, errs chan<- error) {
+			defer wg.Done()
+			for p := range panels {
+				err := rep.renderPNG(p)
+				if err != nil {
+					log.Errorf("Error creating image for panel: %v", err)
+					errs <- err
+				}
+			}
+		}(panels, errs)
+	}
+	wg.Wait()
+	close(errs)
+
+	for err := range errs {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (rep *report) renderPNG(p grafana.Panel) error {
+	body, err := rep.gClient.GetPanelPng(p, rep.dashName, rep.time)
+	if err != nil {
+		return fmt.Errorf("error getting panel %+v: %v", p, err)
+	}
+	defer body.Close()
+
+	err = os.MkdirAll(rep.imgDirPath(), 0777)
+	if err != nil {
+		return fmt.Errorf("error creating img directory:%v", err)
+	}
+	imgFileName := fmt.Sprintf("image%d.png", p.ID)
+	file, err := os.Create(filepath.Join(rep.imgDirPath(), imgFileName))
+	if err != nil {
+		return fmt.Errorf("error creating image file:%v", err)
+	}
+	defer file.Close()
+
+	_, err = io.Copy(file, body)
+	if err != nil {
+		return fmt.Errorf("error copying body to file:%v", err)
+	}
+	return nil
+}
+
+func (rep *report) renderPDF(dash grafana.Dashboard) (outputPDF *os.File, err error) {
+	pdf := gopdf.GoPdf{}
```

Comment:
extract code related to config pdf  to a function and add some comments. some reviewers can review it carelessly